### PR TITLE
FSD-3  Algorithm Concepts

### PR DESCRIPTION
**Pull request by AMubinshaikh**
FSD-3 is based on Algorithm concepts like
Sorting
Searching & recursive
Bubble sort with example of ascending order and descending

QuickSort
Binary Search with and without recursive function

### CODE REVIEW

File: Algorithms_Concepts/Factorial.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        public void DisplayFactorial()
+        {
+            int result;
+            result = CalculateFactorial(number);
+            Console.WriteLine("Factorial = "+ result);
+        }
+        public int CalculateFactorial( int number)
+        {
+           if(number==0 || number==1 )
+           {
```
Comment (abdulmuheet) [inline Algorithms_Concepts/Factorial.cs]: curly braces are nor required for a single line of code in it.

File: Algorithms_Concepts/Factorial.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+            number = Convert.ToInt32(Console.ReadLine());
+            
+        }
+        public void DisplayFactorial()
+        {
+            int result;
+            result = CalculateFactorial(number);
+            Console.WriteLine("Factorial = "+ result);
+        }
+        public int CalculateFactorial( int number)
```
Comment (abdulmuheet) [inline Algorithms_Concepts/Factorial.cs]: remove space before int

File: Algorithms_Concepts/BinarySearch.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+{
+ public class BinarySearch
+ {
+     int arrayLoop;
+        int arraySize;
+        int search;
+        int[] array = new int[100];
+        public void AcceptArray()
+        {
+               Console.WriteLine ("Define Array size");
```
Comment (abdulmuheet) [inline Algorithms_Concepts/BinarySearch.cs]: Instead of getting input from the console, get it in method input parameters.

Comment (abdulmuheet) [reply Algorithms_Concepts/BinarySearch.cs]: This is applicable to all the places.

File: Algorithms_Concepts/BinarySearch.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+               arraySize = Convert.ToInt32 (Console.ReadLine ());
+             
+               for (arrayLoop = 0; arrayLoop < arraySize; arrayLoop++)
+               {
+                   Console.WriteLine ("Enter Value");
+                   array[arrayLoop] = Convert.ToInt32 (Console.ReadLine ());
+               }
+        }
+          public int BinarySearchMethod (int[] arr, int min, int max, int search) {
```
Comment (abdulmuheet) [inline Algorithms_Concepts/BinarySearch.cs]: 'Method' word is not required.

Comment (abdulmuheet) [reply Algorithms_Concepts/BinarySearch.cs]: give better variable name.

File: Algorithms_Concepts/BinarySearchRecursive.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
+/*
```
Comment (abdulmuheet) [inline Algorithms_Concepts/BinarySearchRecursive.cs]: What is the use of this class? all code is commented.

File: Algorithms_Concepts/BubbleSort.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+               for (arrayLoop = 0; arrayLoop < arraySize; arrayLoop++)
+               {
+                   Console.WriteLine ("Enter Value");
+                   array[arrayLoop] = Convert.ToInt32 (Console.ReadLine ());
+               }
+        }
+        public void SortArray()
+        {
+            
+
```
Comment (abdulmuheet) [inline Algorithms_Concepts/BubbleSort.cs]: extra line break not required.

File: Algorithms_Concepts/BubbleSort.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+           for (arrayLoop = 0; arrayLoop < arraySize; arrayLoop++) 
+           {
+              for (int inetrnalLoop = 0; inetrnalLoop < arraySize - 1; inetrnalLoop++)
```
Comment (abdulmuheet) [inline Algorithms_Concepts/BubbleSort.cs]: typo, also rename inetrnalLoop -> 'inner' and the above loop variable to 'outer'

File: Algorithms_Concepts/Factorial.cs [hunk 8]
```diff
@@ -0,0 +1,33 @@
+using System;
+namespace Algorithms_Concepts
+{
+    public class Factorial
+    {
+       public int number;
+        public void AcceptNumber()
+        {
+
```
Comment (abdulmuheet) [inline Algorithms_Concepts/Factorial.cs]: extra line space should be removed from all the places.

File: Algorithms_Concepts/Factorial.cs [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        {
+            
+            Console.WriteLine("Enter a Number");
+            number = Convert.ToInt32(Console.ReadLine());
+            
+        }
+        public void DisplayFactorial()
+        {
+            int result;
+            result = CalculateFactorial(number);
```
Comment (abdulmuheet) [inline Algorithms_Concepts/Factorial.cs]: directly declare the variable here.

File: Algorithms_Concepts/Program.cs [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+             // Recursive Methods Find Element in given Array using Quick Sort
+
+              QuickSort quickSort = new QuickSort();
+              quickSort.AcceptArray();
+              quickSort.QuickSortMethod();
+              quickSort.DisplaySortedArray();
+            
+            //---------------------------
+
+
```
Comment (abdulmuheet) [inline Algorithms_Concepts/Program.cs]: remove extra line spaces.

File: Algorithms_Concepts/QuickSort.cs [hunk 11]
```diff
@@ -0,0 +1,72 @@
+using System;
+namespace Algorithms_Concepts
+{
+    public class QuickSort
+    {
+         int arrayLoop;
```
Comment (abdulmuheet) [inline Algorithms_Concepts/QuickSort.cs]: The indentation is not correct.

File: Algorithms_Concepts/QuickSort.cs [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+            int pivot, leftend, rightend;
+            leftend = left;
+            rightend = right;
+            pivot = array[left];
+            while (left < right) {
+                while ((array[right] >= pivot) && (left < right)) {
+                    right--;
+                }
+
+                if (left != right) {
```
Comment (abdulmuheet) [inline Algorithms_Concepts/QuickSort.cs:21]: add a comment to explain the code wherever some logic is written.

File: Algorithms_Concepts/bin/Debug/netcoreapp3.1/Algorithms_Concepts.deps.json [hunk 13]
```diff
@@ -0,0 +1,23 @@
+{
```
Comment (abdulmuheet) [inline Algorithms_Concepts/bin/Debug/netcoreapp3.1/Algorithms_Concepts.deps.json]: I don't think this file is required to be in source control.

File: Algorithms_Concepts/QuickSort.cs [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+               {
+                   Console.WriteLine ("Enter Value");
+                   array[arrayLoop] = Convert.ToInt32 (Console.ReadLine ());
+               }
+        }
+        public void QuickSortMethod () {
+            sort (0, len - 1);
+        }
+        public void sort (int left, int right) {
+            int pivot, leftend, rightend;
```
Comment (abdulmuheet) [inline Algorithms_Concepts/QuickSort.cs]: use better variable names

File: Algorithms_Concepts/QuickSort.cs [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+            }
+
+            if (right > pivot) {
+                sort (pivot + 1, right);
+            }
+        }
+       
+        public void DisplaySortedArray()
+        {
+            for (int j = 0; j < arraySize; j++)
```
Comment (abdulmuheet) [inline Algorithms_Concepts/QuickSort.cs]: j is not a good name ;)

File: Algorithms_Concepts/Program.cs [hunk 16]
```diff
@@ -0,0 +1,70 @@
+﻿using System;
+
+namespace Algorithms_Concepts
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            //Bubble sot Ascending Order
```
Comment (abdulmuheet) [inline Algorithms_Concepts/Program.cs]: you can uncomment all these codes and pass input array/values in the function.

File: Algorithms_Concepts/BubleSortDescending.cs [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+                 {
+                       int temp = array[inetrnalLoop + 1];
+                        array[inetrnalLoop + 1] = array[inetrnalLoop];
+                        array[inetrnalLoop] = temp;
+                 }
+              }
+            }
+        }
+
+        public void PrintSortArray()
```
Comment (abdulmuheet) [inline Algorithms_Concepts/BubleSortDescending.cs]: Printing should be done on the client side, in program.cs

File: Algorithms_Concepts/BinarySearch.cs [hunk 18]
```diff
@@ -0,0 +1,43 @@
+using System;
+namespace Algorithms_Concepts
+{
+ public class BinarySearch
+ {
+     int arrayLoop;
+        int arraySize;
+        int search;
+        int[] array = new int[100];
```
Comment (abdulmuheet) [inline Algorithms_Concepts/BinarySearch.cs]: rename array -> numbers

**Review by abdulmuheet — CHANGES_REQUESTED** [review]
@AMubinshaikh Please refer to the comment mentioned inline.

File: Algorithms_Concepts/Factorial.cs (hunk 1 above)
Comment (AMubinshaikh) [reply Algorithms_Concepts/Factorial.cs]: done

File: Algorithms_Concepts/BubleSortDescending.cs [hunk 19]
```diff
@@ -0,0 +1,19 @@
+//program for bubble sort     
+//Bubble Sort Descending order
+
+using System;
+namespace Algorithms_Concepts {
+   public class BubbleSortDescending {
+      public void SortDescending (int[] numbers) {
+         for (int arrayLoop = 0; arrayLoop < numbers.Length; arrayLoop++) {
+            for (int inetrnalLoop = 0; inetrnalLoop < numbers.Length - 1; inetrnalLoop++) {
```
Comment (abdulmuheet) [inline Algorithms_Concepts/BubleSortDescending.cs:9]: typo, please correct it.

Comment (abdulmuheet) [reply Algorithms_Concepts/BubleSortDescending.cs:9]: inetrnalLoop it should be internalLoop

**Review by abdulmuheet — APPROVED** [review]
approved

**Pull request merged**